File: test/matlab/test__pid_control__DF.m
%********
%* setup
%********
bin_dir = '../../build/bin';
dat_dir = '../../build/dat';
ref_dat_dir = "../reference_dat";
test_name = 'test-pid_control-DF';
dat_prefix = append(dat_dir, '/', test_name, '-');
ref_dat_prefix = append(ref_dat_dir, '/', test_name, '-');
exe_name = append(test_name, '.exe');
t_arr_fname = 't_arr.dat';
dt__x_arr_fname = 'dt__x_arr.dat';

is_drawing = false;
is_single_precision = false;
f = 3;

if is_single_precision
	error_thres = 1e-1; %* single precision
else 
	error_thres = 1e-1;
end

prev_pwd = pwd;
cd(bin_dir);
if isfile(exe_name)
	%* call the test executable
	if system(exe_name) > 0
		warning(append(bin_dir, '/', exe_name, ' has returned failure.'));
	end
	
	%* read the results
	t_arr = readmatrix(append(dat_prefix, t_arr_fname));
	dt__x_arr = readmatrix(append(dat_prefix, dt__x_arr_fname));

	%* verify
	dt__x_arr_chk = 2*pi*f*cos(t_arr*2*pi*f);

	start_idx = find(t_arr > .1, 1, 'first');
	max_error = max(vecnorm(dt__x_arr(start_idx:end) - dt__x_arr_chk(start_idx:end), 2, 2));
	mean_error = mean(vecnorm(dt__x_arr(start_idx:end) - dt__x_arr_chk(start_idx:end), 2, 2));

	if max_error < error_thres
		disp(append(test_name, '	ok'));
	else
		disp(append(test_name, '	fail'));
	end

	figure('Name', 'x');
	hold on;
	plot(t_arr, dt__x_arr);
	plot(t_arr, dt__x_arr_chk, '--');
	else
	error(append(bin_dir, '/', exe_name, ' does not exist. Use CMake to build the test.'));
end
cd(prev_pwd);

File: test/test-pid_control-DF.cpp
#include "matrix_rw.hpp"
#include "pid_control.hpp"
#include <cmath>

using real_t = pid_control::real_t;
using uint_t = pid_control::uint_t;

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

//* setup
const std::string dat_dir = "../dat";
const std::string ref_dat_dir = "../../test/reference_dat";
const std::string test_name = "test-pid_control-DF";
const std::string dat_prefix = dat_dir + "/" + test_name + "-";
const std::string ref_dat_prefix = ref_dat_dir + "/" + test_name + "-";
const std::string t_arr_fname = "t_arr.dat";
const std::string dt__x_arr_fname = "dt__x_arr.dat";

constexpr uint_t sample_freq = 1e5;
constexpr real_t h = 1. / sample_freq;
constexpr real_t t0 = 0;
constexpr real_t tf = 1;
constexpr uint_t t_dim = sample_freq*(tf - t0) + 1;
constexpr uint_t x_dim = 1;
constexpr real_t sine_freq = 3.;
constexpr real_t T_f = h * 1e1;

#ifdef __USE_SINGLE_PRECISION__
constexpr real_t error_thres = 1e-1;
#else
constexpr real_t error_thres = 1e-1;
#endif

int
main()
{
	//* test
	real_t x[x_dim] = {0};
	real_t dt__x[x_dim] = {0};
	real_t x_next[x_dim];
	real_t t_arr[t_dim];
	real_t dt__x_arr[t_dim * x_dim] = {0};

	for (uint_t i = 0; i < t_dim; ++i) {
		t_arr[i] = i * h;
		x_next[0] = sin(t_arr[i] * 2. * M_PI * sine_freq);

		pid_control::DF<x_dim>(h, T_f, x, x_next, dt__x, dt__x_arr + i);

		//* next x, dt__x becomes previous
		x[0] = x_next[0];
		dt__x[0] = dt__x_arr[i];
	}

	//* write test data
	matrix_rw::write<t_dim, 1>(dat_prefix + t_arr_fname, t_arr);
	matrix_rw::write<t_dim, x_dim>(dat_prefix + dt__x_arr_fname, dt__x_arr);

	//* verify
	real_t max_error = 0.;

	for (uint_t i = 1e4; i < t_dim; ++i) {

		const real_t error = std::abs(dt__x_arr[i] - 2 * M_PI * sine_freq * cos(t_arr[i] * 2 * M_PI * sine_freq));
		if (error > max_error) {
			max_error = error;
		}
	}

	if (max_error < error_thres) {
		return 0;
	} else {
		return 1;
	}
}
